src/pkg/artifact/dao/dao_test.go

```diff
@@ -0,0 +1,201 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dao
+
+import (
+	common_dao "github.com/goharbor/harbor/src/common/dao"
+	ierror "github.com/goharbor/harbor/src/internal/error"
+	"github.com/goharbor/harbor/src/pkg/q"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"testing"
+	"time"
+)
+
+var (
+	typee                   = "IMAGE"
+	mediaType               = "application/vnd.oci.image.config.v1+json"
+	manifestMediaType       = "application/vnd.oci.image.manifest.v1+json"
+	projectID         int64 = 1
+	repositoryID      int64 = 1
+	digest                  = "sha256:418fb88ec412e340cdbef913b8ca1bbe8f9e8dc705f9617414c1f2c8db980180"
+)
+
+type daoTestSuite struct {
+	suite.Suite
+	require    *require.Assertions
+	assert     *assert.Assertions
+	dao        DAO
+	artifactID int64
+}
+
+func (d *daoTestSuite) SetupSuite() {
+	d.dao = New()
+	d.require = require.New(d.T())
+	d.assert = assert.New(d.T())
+	common_dao.PrepareTestForPostgresSQL()
+}
+
+func (d *daoTestSuite) SetupTest() {
+	artifact := &Artifact{
+		Type:              typee,
+		MediaType:         mediaType,
+		ManifestMediaType: manifestMediaType,
+		ProjectID:         projectID,
+		RepositoryID:      repositoryID,
+		Digest:            digest,
+		Size:              1024,
+		PushTime:          time.Now(),
+		PullTime:          time.Now(),
+		ExtraAttrs:        `{"attr1":"value1"}`,
+		Annotations:       `{"anno1":"value1"}`,
+	}
+	id, err := d.dao.Create(nil, artifact)
+	d.require.Nil(err)
+	d.artifactID = id
+}
+
+func (d *daoTestSuite) TearDownTest() {
+	err := d.dao.Delete(nil, d.artifactID)
+	d.require.Nil(err)
+}
+
+func (d *daoTestSuite) TestGetTotal() {
+	// nil query
+	total, err := d.dao.Count(nil, nil)
```

Comment:
You should test the case when there's paging info in the query and verify it doesn't impact the total number.  (100 records and pagesize is 10, the result should be 100)

Reply:
Added